File: FrontEnd/src/app/services/Assurance-service/contrat.service.ts
import { EnvironmentInjector, Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Contrat } from 'src/app/core/models/Contrat.model';

@Injectable({
  providedIn: 'root'
})
export class ContratService {

  private apiUrl = ' http://localhost:8081/api/contrats'; // URI de l'API

  constructor(private http: HttpClient) { }

  // Créer un contrat
  createContrat(contrat: Contrat): Observable<Contrat> {
    return this.http.post<Contrat>(this.apiUrl, contrat);
  }

  // Mettre à jour un contrat
  updateContrat(id: number, contrat: Contrat): Observable<Contrat> {
    return this.http.put<Contrat>(`${this.apiUrl}/${id}`, contrat);
  }

  // Supprimer un contrat
  deleteContrat(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // Obtenir tous les contrats
  getAllContrats(): Observable<Contrat[]> {
    return this.http.get<Contrat[]>(this.apiUrl);
  }

  // Obtenir un contrat par son ID
  getContratById(id: number): Observable<Contrat> {
    return this.http.get<Contrat>(`${this.apiUrl}/${id}`);
  }

  // Créer un contrat à partir d'une assurance
  createContratFromAssurance(assuranceId: number, contrat: Contrat): Observable<Contrat> {
    return this.http.post<Contrat>(`${this.apiUrl}/from-assurance/${assuranceId}`, contrat);
  }

  // Créer un contrat à partir d'un devis
  createContratFromDevis(devisId: number, contrat: Contrat): Observable<Contrat> {
    return this.http.post<Contrat>(`${this.apiUrl}/from-devis/${devisId}`, contrat);
  }
}

File: Microservices/Microservice-Assurance/src/main/java/tn/esprit/Microservice_Assurance/service/ContratService.java
package tn.esprit.Microservice_Assurance.service;

import tn.esprit.Microservice_Assurance.model.Contrat;

import java.util.List;
import java.util.Optional;

public interface ContratService {
    Contrat createContrat(Contrat contrat);
    Contrat updateContrat(Long id, Contrat contrat);
    void deleteContrat(Long id);
    List<Contrat> getAllContrats();
    Optional<Contrat> getContratById(Long id);

    Contrat createContratFromAssurance(Long assuranceId, Contrat contrat);

    Contrat createContratFromDevis(Long devisId, Contrat contrat);

    // Méthode pour affecter un contrat à un utilisateur
}
